File: src/AliasGenerators/Identity/Implementations/Base/IdentityGenerator.cs
//-----------------------------------------------------------------------
// <copyright file="IdentityGenerator.cs" company="lanedirt">
// Copyright (c) lanedirt. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
// </copyright>
//-----------------------------------------------------------------------

namespace AliasGenerators.Identity.Implementations.Base;

using System.Reflection;
using AliasGenerators.Identity;
using AliasGenerators.Identity.Models;

/// <summary>
/// Dutch identity generator which implements IIdentityGenerator and generates
/// random dutch identities.
/// </summary>
public abstract class IdentityGenerator : IIdentityGenerator
{
    /// <summary>
    /// List of male first names in memory.
    /// </summary>
    private readonly List<string> _firstNamesMale;

    /// <summary>
    /// List of female first names in memory.
    /// </summary>
    private readonly List<string> _firstNamesFemale;

    /// <summary>
    /// List of last names in memory.
    /// </summary>
    private readonly List<string> _lastNames;

    /// <summary>
    /// Random instance.
    /// </summary>
    private readonly Random _random = new();

    /// <summary>
    /// Initializes a new instance of the <see cref="IdentityGenerator"/> class.
    /// </summary>
    public IdentityGenerator()
    {
        _firstNamesMale = LoadList(FirstNamesListMale);
        _firstNamesFemale = LoadList(FirstNamesListFemale);
        _lastNames = LoadList(LastNamesList);
    }

    /// <summary>
    /// Gets namespace path to the male first names list for the correct language.
    /// </summary>
    protected virtual string FirstNamesListMale => "AliasGenerators.Identity.Implementations.Dictionaries.nl.firstnames_male";

    /// <summary>
    /// Gets namespace path to the female first names list for the correct language.
    /// </summary>
    protected virtual string FirstNamesListFemale => "AliasGenerators.Identity.Implementations.Dictionaries.nl.firstnames_female";

    /// <summary>
    /// Gets namespace path to the last names list for the correct language.
    /// </summary>
    protected virtual string LastNamesList => "AliasGenerators.Identity.Implementations.Dictionaries.nl.lastnames";

    /// <inheritdoc/>
    public async Task<Identity> GenerateRandomIdentityAsync()
    {
        await Task.Yield(); // Add an await statement to make the method truly asynchronous.

        // Generate identity.
        var identity = new Identity();

        // Determine gender.
        if (_random.Next(2) == 0)
        {
            identity.FirstName = _firstNamesMale[_random.Next(_firstNamesMale.Count)];
            identity.Gender = Gender.Male;
        }
        else
        {
            identity.FirstName = _firstNamesFemale[_random.Next(_firstNamesFemale.Count)];
            identity.Gender = Gender.Female;
        }

        identity.LastName = _lastNames[_random.Next(_lastNames.Count)];

        // Generate random date of birth between 21 and 65 years of age.
        identity.BirthDate = GenerateRandomDateOfBirth();
        identity.EmailPrefix = new UsernameEmailGenerator().GenerateEmailPrefix(identity);
        identity.NickName = new UsernameEmailGenerator().GenerateUsername(identity);

        return identity;
    }

    /// <summary>
    /// Load a list of words from a resource file.
    /// </summary>
    /// <param name="resourceName">Name of the resource file to load.</param>
    /// <returns>List of words from the resource file.</returns>
    /// <exception cref="FileNotFoundException">Thrown if resource file cannot be found.</exception>
    private static List<string> LoadList(string resourceName)
    {
        var assembly = Assembly.GetExecutingAssembly();

        using var stream = assembly.GetManifestResourceStream(resourceName);
        if (stream == null)
        {
            throw new FileNotFoundException("Resource '" + resourceName + "' not found.", resourceName);
        }

        using var reader = new StreamReader(stream);
        var words = new List<string>();
        while (!reader.EndOfStream)
        {
            var line = reader.ReadLine();
            if (line != null)
            {
                words.Add(line);
            }
        }

        return words;
    }

    /// <summary>
    /// Generate a random date of birth.
    /// </summary>
    /// <returns>DateTime representing date of birth.</returns>
    private DateTime GenerateRandomDateOfBirth()
    {
        // Generate random date of birth between 21 and 65 years of age.
        var now = DateTime.Now;
        var minDob = now.AddYears(-65);
        var maxDob = now.AddYears(-21);
        return minDob.AddDays(_random.Next((int)(maxDob - minDob).TotalDays));
    }
}

File: shared/identity-generator/src/implementations/base/IdentityGenerator.ts
import { UsernameEmailGenerator } from '../../utils/UsernameEmailGenerator';
import { Gender } from '../../types/Gender';
import { IIdentityGenerator } from '../../interfaces/IIdentityGenerator';
import { Identity } from '../../types/Identity';

/**
 * Base identity generator.
 */
export abstract class IdentityGenerator implements IIdentityGenerator {
  protected firstNamesMale: string[] = [];
  protected firstNamesFemale: string[] = [];
  protected lastNames: string[] = [];
  private readonly random = Math.random;

  /**
   * Constructor.
   */
  public constructor() {
    // Each implementing class should provide these as static JSON strings
    this.firstNamesMale = this.getFirstNamesMaleJson();
    this.firstNamesFemale = this.getFirstNamesFemaleJson();
    this.lastNames = this.getLastNamesJson();
  }

  protected abstract getFirstNamesMaleJson(): string[];
  protected abstract getFirstNamesFemaleJson(): string[];
  protected abstract getLastNamesJson(): string[];

  /**
   * Generate a random date of birth.
   */
  protected generateRandomDateOfBirth(): Date {
    const today = new Date();
    const minAge = 21;
    const maxAge = 65;

    const minDate = new Date(today.getFullYear() - maxAge, today.getMonth(), today.getDate());
    const maxDate = new Date(today.getFullYear() - minAge, today.getMonth(), today.getDate());

    const timestamp = minDate.getTime() + (this.random() * (maxDate.getTime() - minDate.getTime()));
    return new Date(timestamp);
  }

  /**
   * Generate a random identity.
   */
  public generateRandomIdentity(gender?: string | 'random'): Identity {
    const identity: Identity = {
      firstName: '',
      lastName: '',
      gender: Gender.Male,
      birthDate: new Date(),
      emailPrefix: '',
      nickName: ''
    };

    // Determine gender
    let selectedGender: Gender;
    if (gender === 'random' || gender === undefined) {
      // Random selection (default behavior)
      selectedGender = this.random() < 0.5 ? Gender.Male : Gender.Female;
    } else {
      // Use specified gender
      if (gender === 'male') {
        selectedGender = Gender.Male;
      } else if (gender === 'female') {
        selectedGender = Gender.Female;
      } else {
        selectedGender = Gender.Male;
      }
    }

    // Set gender and appropriate first name
    identity.gender = selectedGender;
    if (selectedGender === Gender.Male) {
      identity.firstName = this.firstNamesMale[Math.floor(this.random() * this.firstNamesMale.length)];
    } else if (selectedGender === Gender.Female) {
      identity.firstName = this.firstNamesFemale[Math.floor(this.random() * this.firstNamesFemale.length)];
    } else {
      // For Gender.Other, randomly choose from either list
      const usesMaleNames = this.random() < 0.5;
      identity.firstName = usesMaleNames
        ? this.firstNamesMale[Math.floor(this.random() * this.firstNamesMale.length)]
        : this.firstNamesFemale[Math.floor(this.random() * this.firstNamesFemale.length)];
    }

    identity.lastName = this.lastNames[Math.floor(this.random() * this.lastNames.length)];
    identity.birthDate = this.generateRandomDateOfBirth();

    const generator = new UsernameEmailGenerator();
    identity.emailPrefix = generator.generateEmailPrefix(identity);
    identity.nickName = generator.generateUsername(identity);

    return identity;
  }
}